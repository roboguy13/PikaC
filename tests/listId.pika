%generate listId

Sll : layout[-x](List);
Sll Nil := emp;
Sll (Cons head tail) :=
  exists +nxt.
  x :-> head **
  (x+1) :-> nxt **
  Sll tail [nxt];

listId : Sll -> Sll;
listId Nil := Nil [Sll];
listId (Cons h t) := Cons h t [Sll];

