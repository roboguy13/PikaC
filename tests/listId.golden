void listId(loc x, loc x1);
void listId(loc x, loc x1) {
 if (x == 0) {
  WRITE_LOC(x1, 0, 0);
 } else {
  loc h = READ_LOC(x, 0);
  loc n = READ_LOC(x, 1);
  loc x = (loc)malloc(2 * sizeof(loc));
  WRITE_LOC(x1, 0, x);
  WRITE_LOC(x, 1, n);
  WRITE_LOC(x, 0, h);
 }
}